### ISSUE TITLE
Semantic Labels for TextSpans with GestureRecognizer do not show up if there are additional TextSpans attached after it

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#66861]
b/169155017

One of the g3 teams have built their own html widget generator which works by emitting several `TextSpan` objects. When one of those objects have a GestureRecognizer (for links), it gets dropped if the text span that comes after it is long enough to not fit into the same row.

```
TextSpan:
   TextSpan:
     "Before link. "
   TextSpan:
     inherit: true
     color: Color(0xff1a73e8)
     semanticsLabel: ""
     TextSpan:
       "This is a link"
       recognizer: TapGestureRecognizer
   TextSpan:
     "After Link, there is some long text that does not fit to a single line."
```

See internal bug for reproduction.

**Comment by chunhtai** [comment]
I cannot reproduce it. This is my current code
```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
      showSemanticsDebugger: true,
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextSpan spans = TextSpan(
    text: 'a',
    children: <InlineSpan> [
      TextSpan(text: 'before link.'),
      TextSpan(
        text: 'b',
        semanticsLabel: '',
        children: <InlineSpan>[
          TextSpan(text: 'this is a link', recognizer: TapGestureRecognizer()..onTap = () => print('link tap!')),
        ]
      ),
      TextSpan(text: 'After Link, there is some long text that does not fit to a single line.')
    ]
  );

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(44.0),
      child: _buildHtmlText(context),
    );
  }

  Widget _buildHtmlText(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(44.0),
      child: Text.rich(spans),
    );
  }
}
```

The repro in the internal bug uses custom widgets and I cannot import to my local app.

**Comment by mehmetf** [comment]
I can reproduce this.

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
      showSemanticsDebugger: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextSpan spans = TextSpan(
    text: 'a',
    children: <InlineSpan> [
      TextSpan(text: '\n\n\n\n'),
      TextSpan(
        text: 'b',
        semanticsLabel: '',
        children: <InlineSpan>[
          TextSpan(text: 'This is a link', recognizer: TapGestureRecognizer()..onTap = () => print('link tap!')),
        ]
      ),
      TextSpan(text: '\n\n\n\n'),
    ]
  );

  @override
  Widget build(BuildContext context) {
    return 
     Scaffold(
        appBar: AppBar(
          title: Text('Welcome to Flutter'),
        ),
        body: Container(
          padding: EdgeInsets.all(44.0),
          child: _buildHtmlText(context),
        ),
      );
  }

  Widget _buildHtmlText(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(44.0),
      child: Text.rich(spans),
    );
  }
}
```

This is with semantic debugger off. You will see this:

![image](https://user-images.githubusercontent.com/6687929/94621113-997ca500-0264-11eb-9b8d-fea191230a58.png)

Turning debugger on, you will see:

![image](https://user-images.githubusercontent.com/6687929/94621157-ac8f7500-0264-11eb-9e04-e7e86f159e50.png)

**Comment by mehmetf** [comment]
The bottom line's semantics box completely obscures the link.

**Comment by chunhtai** [comment]
from a quick glance, it is probably because the flutter does not know how to handle non rect boundary.

The correct node rect should be as the following map
![Untitled drawing](https://user-images.githubusercontent.com/47866232/94855523-82b08c80-03e3-11eb-94d4-45a544c8bde9.png)

but this is not possible because android accessibility(probably ios too) only accept rectangular node.

I try to repro with a native android app, but it seems either an entire paragraph is treated as a node, or you can separate into sub paragraph, but then you won't be able to have them to wrap like a single paragraph.

**Comment by mehmetf** [comment]
How does Android native handle this for webviews? 

The main issue seems to be links so a blob of text interspersed with links on a webpage would be a good baseline.

**Comment by chunhtai** [comment]
![webview](https://user-images.githubusercontent.com/47866232/94861415-577e6b00-03ec-11eb-84ba-6da22203564f.gif)

I believe what we have currently matches the behavior for the web too. it will be covered, but it is reachable through swipe

**Comment by mehmetf** [comment]
Thanks @chunhtai . I will close this as WAI.
